harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -636,3 +654,36 @@ def test_checkpoint_loading(ckpt: str, istrial: bool):
         assert isinstance(trial, pytorch.PyTorchTrial), type(trial)
     else:
         assert isinstance(trial, torch.nn.Module), type(trial)
+
+
+def make_amp_workloads(
+    trial_class,
+    assert_output_float16=False,
+) -> workload.Stream:
+    trainer = utils.TrainAndValidate()
+    yield from trainer.send(steps=10, validation_freq=1, scheduling_unit=1)
+    training_metrics, _ = trainer.result()
+
+    scale_ever_decreased = False
+    scale_ever_increased = False
+    for idx, (older, newer) in enumerate(zip(training_metrics, training_metrics[1:])):
+        if assert_output_float16:
+            assert newer["output"].dtype is np.dtype("float16")
+        else:
+            # Automatic usages of Apex or native AMP cast the output back to float32
+            # For the latter case, see the hook end_f16
+            #   defined in PyTorchTrialContext.autocast_forward_pass
+            assert newer["output"].dtype is np.dtype("float32")
+        if newer["scale"] != older["scale"]:
+            scale_ever_decreased = scale_ever_decreased or newer["scale"] < older["scale"]
+            scale_ever_increased = scale_ever_increased or newer["scale"] > older["scale"]
+            # If the scale changed, it's because an inf or NaN was encountered, so the
+            #  model couldn't have changed this iteration.
+            continue
+        assert newer["loss"] <= older["loss"]
+        assert newer["loss"].dtype is np.dtype("float32")
+        # Check the accuracy of the gradient change.
+        trial_class.check_batch_metrics(newer, idx, epsilon=1e-3)
+        if idx == 0:
+            trial_class.check_batch_metrics(older, idx, epsilon=1e-3)
+    assert scale_ever_decreased or scale_ever_increased
```

Comment:
While this code works and does what it's supposed to, it could be tightened up quite a bit.

Reply:
If you think there's a lot of tightening up to do, now is the time to do it.

I'd be interested to be a bit more precise about when the scale overflows or underflows.  Right now, which batch index is the underflow/overflow it happening on?  The OnesDataset seems to me like it's going to have a pretty consistent and manageable gradient.  I thought you might change the dataset to emit some particular record of 100000 or something to force it to overflow just when you wanted it to.

Reply:
The scaler factor is initiailly 2e16 by default, so the ones data is already "large", which causes the rescaling and the skipping of the optimizer. If you were to set, say, `init_scale=8` in the `GradScaler` constructor and rerun the manual test it would fail.

Reply:
I'd still like to see the test be a little more precise about what the exact losses are for every batch.

But perhaps, my request might need a bit of explanation as to what is coming down the line for our PyTorchTrial unit tests.

Eventually I want unit tests for (gradient aggregation + amp), (checkpointing + amp), and (dtrain + amp).  Those tests will need to be very precise about what exact losses and scale factors are expected at each batch index.  I'm thinking of a function like:

```python
def expected_loss_scale_weights(num_batches):
    out = []
    for i in range(num_batches):
        loss = ...
        scale = ...
        weight = ...
        out.append((loss, scale, weight))
    return out
```

In the current test, this is dangerously close to just testing AMP, but if you mix amp with dtrain, gradient aggregation, or checkpointing, where we have our hands all over various states, having an source of truth which is absolute (based on mathematical principles, not based on the previous output of the train_batch function) will mean that the one source of truth works for all of those tests.

Also, from a purely testing perspective, I trust test-against-ground-truth much more than test-by-inductive-reasoning.  I told the same thing to Maksim when he was writing the `test_gradient_aggregation` test.

Reply:
^ i think we'll also need this to verify the path where a loss underflow/overflows for a particular batch and the scale changes but the loss doesn't step. not sure if you're planning on adding that in this PR, but we should probably check for that in a comprehensive AMP test.